Add libc_bitflags convenience macro

kamalmarhubi opened the pull request:
We define many bitflags types with values from the libc crate. Currently
these look like this:

    bitflags!{
        flags ProtFlags: libc::c_int {
            const PROT_NONE      = libc::PROT_NONE,
            const PROT_READ      = libc::PROT_READ,
            const PROT_WRITE     = libc::PROT_WRITE,
            const PROT_EXEC      = libc::PROT_EXEC,
            #[cfg(any(target_os = "linux", target_os = "android"))]
            const PROT_GROWSDOWN = libc::PROT_GROWSDOWN,
            #[cfg(any(target_os = "linux", target_os = "android"))]
            const PROT_GROWSUP   = libc::PROT_GROWSUP,
        }
    }

There's some repetition which is tedious. With the new macro, the above
can instead be written

    libc_bitflags!{
        flags ProtFlags: libc::c_int {
            PROT_NONE,
            PROT_READ,
            PROT_WRITE,
            PROT_EXEC,
            #[cfg(any(target_os = "linux", target_os = "android"))]
            PROT_GROWSDOWN,
            #[cfg(any(target_os = "linux", target_os = "android"))]
            PROT_GROWSUP,
        }
    }

Thanks to Daniel Keep for the Little Book of Rust Macros, and for
helping with this macro.

Refs https://github.com/nix-rust/nix/issues/264

kamalmarhubi commented:
r? @fiveop @posborne

fiveop commented on src/macros.rs:
````diff
@@ -0,0 +1,194 @@
+///
+/// # Example
+/// ```
+/// # #[macro_use] extern crate nix;
````
I would not make this publicly available.

kamalmarhubi replied:
ah this was me misunderstanding how macro visibility worked. good catch!

fiveop commented:
I like this change.

Can you change all occurences instead of just one, right away. It should not be so many.

kamalmarhubi commented:
Going through the rest now.

posborne commented:
Got my approval as well, although I have not gone through the macros with a fine-tooth comb (but agreed with @fiveop on mot making the macro public).  I would be fine with merging this even if we don't have 100% of libc flags converted.

Once you are happy with things, r=me is fine.

kamalmarhubi commented:
I changed visibility of the macro, and added the use of the macro to the conventions doc.

> I would be fine with merging this even if we don't have 100% of libc flags converted.

Having spent most of the morning today trying to convert ~60 uses with medium success, I'd much rather merge this and do the rest incrementally. The conversion should happen but isn't high priority.

Leaving this to one of you to r+ taking into account the conventions change.

posborne commented:
Looks good to me.  @homu r+

homu commented:
:pushpin: Commit c9e5a15 has been approved by `posborne`

<!-- @homu r=posborne c9e5a15bc381e6f0568e7d2b5b41c8bf46a1844d -->

homu commented:
:hourglass: Testing commit c9e5a15 with merge 4715dde...

kamalmarhubi commented:
@homu retry 7038da0

kamalmarhubi commented:
@homu retry

kamalmarhubi commented:
One day I'll get the hang of @homu r+ 7038da0

homu commented:
:hourglass: Testing commit 7038da0 with merge 431e470...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/nix-rust/nix/builds/116519052)

kamalmarhubi commented:
Oh I guess that should have been r=@posborne, ah well
